compiler/noirc_evaluator/src/ssa/opt/loop_invariant.rs

```diff
@@ -231,6 +232,31 @@ impl<'f> LoopInvariantContext<'f> {
                     false
                 }
             }
+            Instruction::Binary(binary) => {
+                if !matches!(binary.operator, BinaryOp::Add | BinaryOp::Mul) {
+                    return false;
+                }
+
+                let operand_type =
+                    self.inserter.function.dfg.type_of_value(binary.lhs).unwrap_numeric();
+
+                let lhs_const =
+                    self.inserter.function.dfg.get_numeric_constant_with_type(binary.lhs);
+                let rhs_const =
+                    self.inserter.function.dfg.get_numeric_constant_with_type(binary.rhs);
+                let (lhs, rhs) = match (
+                    lhs_const,
+                    rhs_const,
+                    self.outer_induction_variables.get(&binary.lhs),
+                    self.outer_induction_variables.get(&binary.rhs),
+                ) {
+                    (Some((lhs, _)), None, None, Some(upper_bound)) => (lhs, *upper_bound),
+                    (None, Some((rhs, _)), Some(upper_bound), None) => (*upper_bound, rhs),
```

Comment:
Do we care that these middle cases are None or should any of them be match-all cases?

Reply:
I'm not totally sure I understand the question. Could you elaborate? If we have `None` it just means we do not have a constant or an outer induction variable. We only care about the cases where if we have a constant, that the other side of the binary op is an induction variable from the outer loop.

Reply:
I was somewhat trying to understand if we needed no constant/outer induction variable or if using a wildcard match like `(Some((lhs, _)), _, _, Some(upper_bound))` would also be applicable here

Reply:
Ah I see. I could use the wildcard. If we have something like `(Some((lhs, _)), _, _, Some(upper_bound))` it should be impossible for rhs_const to return anything other than `None` as that binary op should already be simplified. Same with fetching the induction variable for `binary.lhs`. If we have a constant lhs, that value should not also be an induction variable. I was thinking the explicit `None` makes this logic a bit clearer to understand, but I am also fine switching to the wildcard syntax.

Reply:
Ah, no that's alright. If it's clearer without the wildcard / is expected to always be None that's fine. I was just trying to understand this logic if it needed / was expected to be None or if there could ever be Some cases and what that'd mean in this context.